### ISSUE TITLE
Setup CD conversion support

### ISSUE THREAD

**Issue by TheJJ** [SFTtech/openage#82]
Very related to the conversion targets (#79) and the version detection (#61) is the support of converting media and game files from setup disk images. The Python module `openage.cabextract` provides cab extraction features needed by the `.cab` files on the disc.
1. Pass mount point of the original CD (or ISO image)
2. Detect version of the disk image
3. Use `fslike` abstraction to mount all archives on disk.
4. Convert contents of mounted archives.

The invocation should originate from `openage/convert/driver.py`.

**Comment by caschulz88** [comment]
Did anyone already implemented something related to this issue in the past? Otherwise I would like to start doing so because I don't really want to install WINE just to extract the media files from the AoE2 installer.

Do we have CD file layouts from all editions that are available out there? I think they might differ from version to version. I only have the first release fullversion on CD-ROM available here. Would be great to get information on differences between this release and others like the Gold Edition or the extension pack.

**Comment by TheJJ** [comment]
We have a cab reader and filesystem abstraction so you can transparently access all the archives on the CD. Mounting the iso should be done through `udisksctl` or something.

You can find the cab reading in `openage/cabextract/`.
The version detection is in `openage/convert/driver.py`, you might wanna improve and extend it.

We require the 1.0c patch i think, as the data format of `empires.dat` might be different before that.

* Mount the CD via `udiskctl`
* Assemble the file structure via `util.fslike`
* Adjust the converter to do better version detection and extract whatever is needed (`driver.py`)

**Comment by azrdev** [comment]
@caschulz88 I collected some of the disk layouts here: <https://gist.github.com/azrdev/e20c13f360840bc0157c>. No guarantees about those without "verified original" in the title, though.
You can fork that gist as any other repo, maybe even to PRs. I'm happy to extend the list!

---
w.r.t. Implementing disk support: We have to decide if GameVersion should be different for an install disk and its installed variant. Since an installation (say, AoK+TC) often maps to multiple different disks.

I tend to make GameVersion an abstract class, whose singleton subclasses know how to identify themselves on a given sourcedir, and that known where the files the conversion needs are (see e.g. the `empires2_x1` being named `EMIRE_1.DAT` in AOE2_GOLD).
The Disks would then have separate subclasses, probably.

Things might get even more complicated if we need multiple disks: AOE2 disk on its own lacks the TC content, but does TC disk have everything we need, or do we require both disks (how would we handle that)? AFAIK 1.0c patch is a requirement, what would we do with a TC disk pre-1.0c (does such exist?)

**Comment by zuntrax** [comment]
Yup, pretty sure pre 1.0c TC disks exists. The patch can be downloaded from the internet. The `Age2XPatch.exe` contains a cab archive at offset 0x4ADC8. This archive seems to  be compressed with MSZIP (aka deflate aka rfc1951). We should implement `MSZIPStream` similar to `LZXDStream` to get this working. `import zlib` is our friend.
